### PR TITLE
RFC: Ignore NaNs on 0.6 (like on 0.5): fixes #796

### LINKED ISSUE #796: xlim and ylim defaults are incorrect on 0.6
On 0.5, the defaults for `xlim` and `ylim` take into account all elements of the plot and try to accommodate them all; on 0.6, they seem to default to (0, 1).

### PR DESCRIPTION
**Pull request by mkborregaard**
Bases summary functions on NaNMath:
min, max, extrema, minimum, maximum

### CODE REVIEW

File: src/Plots.jl [hunk 1]
```diff
@@ -106,6 +106,15 @@ export
 
 # ---------------------------------------------------------
 
+import NaNMath
+# define functions (e.g. `_extrema`, that ignores NaNs, when the type is applicable. To overcome the destructive effects of https://github.com/JuliaLang/julia/pull/12563
+for fun in (:extrema, :minimum, :maximum)
```
Comment (mkborregaard) [inline src/Plots.jl]: An alternative to the eval magic would be to just write it out - less concise, but probably easier for people wondering where `_minimum` is defined and what it does. Thoughts?

**Comment by mkborregaard** [comment]
Marked WIP because this awaits a release of NaNMath

File: src/Plots.jl [hunk 2]
```diff
@@ -106,6 +106,15 @@ export
...
+       @eval $(Symbol(string("_",fun))){F<:AbstractFloat}(x::AbstractArray{F}) = NaNMath.$(fun)(x)
```
Comment (dpsanders) [inline src/Plots.jl]: You don't need `string` here:
```jl
julia> f = :hello
julia> Symbol("_", f)
:_hello
```

File: src/Plots.jl (hunk 1 above)
Comment (dpsanders) [inline src/Plots.jl]: Probably preferable to write it out explicitly.

Comment (mkborregaard) [inline src/Plots.jl]: I know, right? It just feels like you should use the classic metaprogramming example when there is a chance :-) Oh well.

Comment (daschw) [inline src/Plots.jl]: Just out of curiosity: Why didn't you define a `_min` and `_max` here?

**Comment by mkborregaard** [comment]
Great! I didn't define those because the min and max functions in NaNMath are compliant will all of the types of the Base functions, so no reason to dispatch via a local function in Plots. But I am actually right now looking into removing it for the other functions too, so it calls the NaNMath versions explicitly where applicable. It'll require a bit more review, but be more transparent and easier to maintain, I think.

File: src/backends/glvisualize.jl [hunk 3]
```diff
@@ -1382,7 +1382,7 @@ end
 
 
 function gl_heatmap(x,y,z, kw_args)
-    get!(kw_args, :color_norm, Vec2f0(extrema(z)))
+    get!(kw_args, :color_norm, Vec2f0(NaNMath.extrema(z)))
```
Comment (mkborregaard) [inline src/backends/glvisualize.jl]: @SimonDanisch I have replaced `minimum`, `mean`, `maximum`, `max` and `min` with versions from the NaNMath package that ignore `NaN`, to maintain the julia-0.5 behaviour of these functions. I have only done the changes where I believe the argument is likely to be a subtype of `AbstractFloat` or `AbstractArray{<:AbstractFloat}` as I believe this is the most likely place to find `NaN`s. Can you check that 1. all of the changes appear valid, and 2. that we don't need to identify NaN-compliant versions for other types? E.g. in line 1428 you take the `minimum` of an `AABB` object, but I'm guessing that will never be NaN?

File: src/backends/gr.jl [hunk 4]
```diff
@@ -824,7 +824,7 @@ function gr_display(sp::Subplot{GRBackend}, w, h, viewport_canvas)
             # create the colorbar of contour levels
             if sp[:colorbar] != :none
                 gr_set_viewport_cmap(sp)
-                l = round(Int32, 1000 + (h - minimum(h)) / (maximum(h) - minimum(h)) * 255)
+                l = round(Int32, 1000 + (h - NaNMath.minimum(h)) / (NaNMath.maximum(h) - NaNMath.minimum(h)) * 255)
```
Comment (mkborregaard) [inline src/backends/gr.jl]: @SimonDanisch I have replaced minimum, mean, maximum, max and min with versions from the NaNMath package that ignore NaN, to maintain the julia-0.5 behaviour of these functions. I have only done the changes where I believe the argument is likely to be a subtype of AbstractFloat or AbstractArray{<:AbstractFloat} as I believe this is the most likely place to find NaNs. Can you check that 1. all of the changes appear valid, and 2. that we don't need to identify NaN-compliant versions for other types? I am particular uncertain about this line.

File: src/components.jl [hunk 5]
```diff
@@ -645,8 +645,8 @@ function (bc::BezierCurve)(t::Real)
     p
 end
 
-Base.mean(x::Real, y::Real) = 0.5*(x+y)
-Base.mean{N,T<:Real}(ps::FixedSizeArrays.Vec{N,T}...) = sum(ps) / length(ps)
+# mean(x::Real, y::Real) = 0.5*(x+y) #commented out as I cannot see this used anywhere and it overwrites a Base method with different functionality
```
Comment (mkborregaard) [inline src/components.jl]: @tbreloff I am removing two functions here that 1. overwrite Base types, 2. are a little hassle to define NaN-compliant functions for, and 3. don't appear to be used anywhere in the code.

File: src/recipes.jl [hunk 6]
```diff
@@ -550,7 +550,7 @@ Plots.@deps stepbins path
 function _auto_binning_nbins{N}(vs::NTuple{N,AbstractVector}, dim::Integer; mode::Symbol = :auto)
     _cl(x) = max(ceil(Int, x), 1)
```
Comment (mkborregaard) [inline src/recipes.jl]: @oschulz the `ceil(Int, x)` but will fail if x is ever NaN. I assume this is not a problem? (just checking)

Comment (oschulz) [inline src/recipes.jl]: Hm, maybe we better do

    _cl(x) = ceil(Int, max(x, one(x)))

instead, just in case. Well spotted!

Comment (mkborregaard) [inline src/recipes.jl]: So, on 0.6 we'll have to do `_cl(x) = ceil(Int, NaNMath.max(x, one(x)))` - that is the whole point of this PR :-D
I'll put in the change.

Comment (oschulz) [inline src/recipes.jl]: Sounds good!

File: src/backends/gr.jl (hunk 4 above)
Comment (jheinen) [inline src/backends/gr.jl]: I checked the GR backend code and I don't see any reason why not use the NaN-compliant functions. You can safely replace them. There is another call in line 642:

```julia
rmin, rmax = GR.adjustrange(minimum(r), maximum(r))
```

Comment (mkborregaard) [inline src/backends/gr.jl]: Thank you!

**Comment by mkborregaard** [comment]
I am going to go ahead and merge this for some forward progress towards the 0.6 version.

**Comment by dpsanders** [comment]
Thanks for your great work on this!

**Comment by mkborregaard** [comment]
Thanks a lot for the help!

**Pull request merged**